src/Containers/Microsoft.NET.Build.Containers/Registry.cs

```diff
@@ -339,68 +343,186 @@ public async Task<string> DownloadBlobAsync(string repository, Descriptor descri
+
+    /// <summary>
+    /// Upload a blob to the registry using chunks. The chunks are uploaded sequentially, and the size of the chunks is determined by the UploadInformation derived from the 'Start' call
+    /// </summary>
+    private async Task<UploadFinalizeInformation> UploadBlobChunkedAsync(string repository, Stream contents, HttpClient client, UploadInformation uploadInfo, Action<LogMessage> logProgressMessage, CancellationToken cancellationToken)
     {
         cancellationToken.ThrowIfCancellationRequested();
-        Uri patchUri = uploadUri.Uri;
-        var localUploadUri = new UriBuilder(uploadUri.Uri);
-        localUploadUri.Query += $"&digest={Uri.EscapeDataString(digest)}";
+        Uri patchUri = uploadInfo.uploadUri.Uri;
+        contents.Seek(0, SeekOrigin.Begin);
 
-        // TODO: this chunking is super tiny and probably not necessary; what does the docker client do
-        //       and can we be smarter?
-
-        byte[] chunkBackingStore = new byte[MaxChunkSizeBytes];
+        using IncrementalHash hash = IncrementalHash.CreateHash(HashAlgorithmName.SHA256);
+        byte[] chunkBackingStore = new byte[uploadInfo.EffectiveChunkSize];
 
         int chunkCount = 0;
         int chunkStart = 0;
+        // chunked upload retries inspired from regclient -
+        // we allow up to N retries across the whole operation, and
+        // successes 'reclaim' retry attempts for future failures.
+        int retryMax = 10;
+        int retryCount = 0;
 
         while (contents.Position < contents.Length)
         {
             cancellationToken.ThrowIfCancellationRequested();
-            int bytesRead = await contents.ReadAsync(chunkBackingStore, cancellationToken).ConfigureAwait(false);
+            var (content, bytesRead) = await ReadChunk();
+            HttpResponseMessage patchResponse = await client.PatchAsync(patchUri, content, cancellationToken).ConfigureAwait(false);
 
-            ByteArrayContent content = new (chunkBackingStore, offset: 0, count: bytesRead);
-            content.Headers.ContentType = new MediaTypeHeaderValue("application/octet-stream");
-            content.Headers.ContentLength = bytesRead;
+            // central state machine -
+            // successes move to the next chunk,
+            // failures are retried up to retryMax times,
+            // or else we fail the whole operation.
+            (patchResponse.StatusCode switch
+            {
+                HttpStatusCode.Accepted => ProcessSuccessfulChunk(patchResponse),
+                // known quirk - AWS ECR sends 201 Created instead of 202 Accepted
+                HttpStatusCode.Created when IsAmazonECRRegistry => ProcessSuccessfulChunk(patchResponse, ignoreRange: true),
+                _ when retryCount < retryMax => await CheckReadAmountAndRetry(patchUri, cancellationToken).ConfigureAwait(false),
+                _ => await LogError(patchResponse, patchUri, cancellationToken).ConfigureAwait(false)
+            })();
```

Comment:
Please excuse this odd usage of switch expressions - they cannot currently return void, so all branches are returning `Action` and I'm immediately invoking them. This is a pretty widely-used hack around current C# language limitations on switch expressions. Yes, I've complained to Kathleen :)